session.go

```diff
@@ -3827,19 +3827,20 @@ func (iter *Iter) Timeout() bool {
 func (iter *Iter) Next(result interface{}) bool {
 	iter.m.Lock()
 	iter.timedout = false
-	timeout := time.Time{}
+	deadline := time.Time{}
 
 	// check should we expect more data.
 	for iter.err == nil && iter.docData.Len() == 0 && (iter.docsToReceive > 0 || iter.op.cursorId != 0) {
-		// we should expect more data.
+		// either there is a getMore running or we will schedule one, set
+		// the deadline regardless.
+		if deadline.IsZero() {
```

Comment:
There's subtle interplay between having deadline and having iter.timeout.  I suggest protecting this if block with `deadline.IsZero && iter.timeout >= 0`, then replace line 3842 with `if ! deadline.IsZero`.